File: backend/tools/allowed_tools.py
"""Comprehensive list of allowed tools for autonomous agents"""

ALLOWED_TOOLS = {
    "network_recon": {"nmap", "rustscan", "masscan", "zmap", "naabu", "scapy", "hping3", "nping", "arp-scan", "netdiscover", "fping", "dnsrecon", "dnsenum", "dnsmap", "dnswalk", "altdns", "amass", "subfinder", "assetfinder", "findomain", "httprobe", "httpx", "waybackurls", "gau", "gospider", "katana", "dnsx", "massdns", "puredns", "shuffledns", "whois", "dig", "nslookup"},
    "web_scanning": {"nikto", "sqlmap", "dirb", "gobuster", "wfuzz", "ffuf", "aquatone", "eyewitness", "gowitness", "whatweb", "wappalyzer-cli", "cmseek", "cmsmap", "droopescan", "joomscan", "wpscan", "drupwn", "curl", "wget", "burpsuite", "zaproxy", "nuclei", "wafw00f"},
    "vuln_scanning": {"trivy", "grype", "snyk", "sonarqube", "clair", "anchore", "semgrep", "checkov", "terrascan", "tfsec", "openscap", "lynis"},
    "exploitation": {"metasploit", "msfvenom", "searchsploit", "hashcat", "john", "hydra", "medusa", "ncrack"},
    "cloud_security": {"aws-cli", "pacu", "prowler", "cloudmapper", "az", "gcloud", "gsutil", "checkov", "cloudsplaining"},
    "active_directory": {"crackmapexec", "impacket", "bloodhound", "sharphound", "pypykatz", "mimikatz", "secretsdump", "evil-winrm", "kerbrute"},
    "container": {"docker", "kubectl", "kubeadm", "kube-bench", "kube-hunter", "trivy", "cdk", "peirates"},
    "dev_tools": {"git", "python", "node", "java", "gcc", "make", "vim", "nano", "sed", "awk", "grep"},
    "system_info": {"uname", "whoami", "hostname", "ifconfig", "ip", "netstat", "ps", "top", "lsof", "id"},
    "database": {"sqlite3", "mysql", "postgresql", "mongosh", "redis-cli"},
    "forensics": {"volatility", "binwalk", "strings", "file", "exiftool", "sleuthkit", "autopsy"},
    "wireless": {"aircrack-ng", "wireshark", "tcpdump", "kismet", "hashcat"},
    "reverse_engineering": {"ghidra", "objdump", "strings", "readelf", "radare2", "r2"},
    "malware_analysis": {"cuckoo", "yara", "strings", "volatility", "wireshark"},
    "osint": {"recon-ng", "theHarvester", "shodan", "censys", "whois"},
    "payloads": {"msfvenom", "donut", "weevely", "netcat", "nc"},
    "api_security": {"postman", "burpsuite", "zaproxy", "owasp-zap"},
}

ALL_ALLOWED_TOOLS = set()
for tools in ALLOWED_TOOLS.values():
    ALL_ALLOWED_TOOLS.update(tools)

def is_tool_allowed(cmd: str) -> bool:
    tool = cmd.split()[0] if cmd else ""
    if tool.startswith("RUN "):
        tool = tool[4:].split()[0]
    return tool.lower() in ALL_ALLOWED_TOOLS

def is_dangerous_command(cmd: str) -> bool:
    dangerous = {"rm -rf", "mkfs", "chmod 777", "reboot", "shutdown", "halt", ":(){:|:&};:"}
    return any(pattern.lower() in cmd.lower() for pattern in dangerous)

def get_allowed_tools_by_category() -> dict:
    return ALLOWED_TOOLS

def get_tool_category(tool: str) -> str:
    for cat, tools in ALLOWED_TOOLS.items():
        if tool.lower() in tools:
            return cat
    return "unknown"

File: backend-go/tools/allowed_tools.go
package tools

var AllowedTools = map[string][]string{
	"network_recon": {
		"nmap", "rustscan", "masscan", "naabu", "dnsrecon", "dnsenum",
		"amass", "subfinder", "httpx", "whois", "dig", "nslookup",
		"fping", "arp-scan", "netdiscover", "dnsx", "massdns",
	},
	"web_scanning": {
		"nikto", "sqlmap", "gobuster", "ffuf", "nuclei", "whatweb",
		"wpscan", "curl", "wget", "dirb", "wfuzz", "wafw00f",
	},
	"vuln_scanning": {
		"trivy", "grype", "semgrep", "lynis", "openscap", "snyk",
	},
	"exploitation": {
		"metasploit", "msfvenom", "searchsploit", "hashcat", "john",
		"hydra", "medusa", "ncrack",
	},
	"cloud_security": {
		"aws-cli", "pacu", "prowler", "az", "gcloud", "checkov",
	},
	"container": {
		"docker", "kubectl", "kube-bench", "kube-hunter", "trivy",
	},
	"osint": {
		"recon-ng", "theHarvester", "shodan", "censys", "maltego",
	},
	"system_info": {
		"uname", "whoami", "hostname", "ifconfig", "ip", "netstat", "ps",
		"top", "lsof", "id", "cat", "ls", "find", "grep",
	},
	"database": {
		"sqlite3", "mysql", "psql", "mongosh", "redis-cli",
	},
	"forensics": {
		"volatility", "binwalk", "strings", "file", "exiftool",
	},
	"wireless": {
		"aircrack-ng", "wireshark", "tcpdump", "kismet",
	},
	"api_security": {
		"postman", "burpsuite", "zaproxy", "owasp-zap",
	},
}

var DangerousCommands = []string{
	"rm -rf", "mkfs", "chmod 777", ":(){:|:&};:",
	"reboot", "shutdown", "halt", "dd if=/dev/zero",
}

func GetAllAllowedTools() []string {
	var all []string
	for _, tools := range AllowedTools {
		all = append(all, tools...)
	}
	return all
}

func IsToolAllowed(tool string, requestedTools []string, allowedToolsOnly bool) bool {
	if !allowedToolsOnly || len(requestedTools) == 0 {
		return isInAllowedTools(tool)
	}
	
	for _, t := range requestedTools {
		if t == tool {
			return true
		}
	}
	return false
}

func isInAllowedTools(tool string) bool {
	for _, tools := range AllowedTools {
		for _, t := range tools {
			if t == tool {
				return true
			}
		}
	}
	return false
}

func IsDangerousCommand(cmd string) bool {
	for _, dangerous := range DangerousCommands {
		if contains(cmd, dangerous) {
			return true
		}
	}
	return false
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(s) > 0 && containsSubstring(s, substr))
}

func containsSubstring(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

func GetToolCategory(tool string) string {
	for category, tools := range AllowedTools {
		for _, t := range tools {
			if t == tool {
				return category
			}
		}
	}
	return "unknown"
}

func FilterToolsByCategory(category string) []string {
	if tools, exists := AllowedTools[category]; exists {
		return tools
	}
	return []string{}
}
